### PR TITLE
feat!: make mapper more strict and allow flexible mode

### PR DESCRIPTION
**Pull request by romm**
The mapper is now more type-sensitive and will fail in the following
situations:

- When a value does not match exactly the awaited scalar type, for
  instance a string `"42"` given to a node that awaits an integer.

- When unnecessary array keys are present, for instance mapping an array
  `['foo' => …, 'bar' => …, 'baz' => …]` to an object that needs only
   `foo` and `bar`.

- When permissive types like `mixed` or `object` are encountered.

These limitations can be bypassed by enabling the flexible mode:

```php
(new \CuyZ\Valinor\MapperBuilder())
    ->flexible()
    ->mapper();
    ->map('array{foo: int, bar: bool}', [
        'foo' => '42', // Will be cast from `string` to `int`
        'bar' => 'true', // Will be cast from `string` to `bool`
        'baz' => '…', // Will be ignored
    ]);
```

When using this library for a provider application — for instance an API
endpoint that can be called with a JSON payload — it is recommended to
use the strict mode. This ensures that the consumers of the API provide
the exact awaited data structure, and prevents unknown values to be
passed.

When using this library as a consumer of an external source, it can make
sense to enable the flexible mode. This allows for instance to convert
string numeric values to integers or to ignore data that is present in
the source but not needed in the application.

---

All these changes led to a new check that runs on all registered object
constructors. If a collision is found between several constructors that
have the same signature (the same parameter names), an exception will be
thrown.

```php
final class SomeClass
{
    public static function constructorA(string $foo, string $bar): self
    {
        // …
    }

    public static function constructorB(string $foo, string $bar): self
    {
        // …
    }
}

(new \CuyZ\Valinor\MapperBuilder())
    ->registerConstructor(
        SomeClass::constructorA(...),
        SomeClass::constructorB(...),
    )
    ->mapper();
    ->map(SomeClass::class, [
        'foo' => 'foo',
        'bar' => 'bar',
    ]);

// Exception: A collision was detected […]
```

### CODE REVIEW

**Comment by eigan** [comment]
I left some comments [here](https://github.com/CuyZ/Valinor/issues/114#issuecomment-1159157941) and [here](https://github.com/CuyZ/Valinor/issues/148#issuecomment-1159170242).

File: docs/pages/mapping/type-strictness.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# Type strictness
+
+The mapper is sensitive to the types of the data that is recursively populated —
+for instance a string `"42"` given to a node that awaits an integer will make
```
Comment (Mopolo) [inline docs/pages/mapping/type-strictness.md:4]:
```suggestion
for instance a string `"42"` given to a node that expects an integer will make
```

File: src/Mapper/Object/Exception/UnexpectedArrayKeysForClass.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    /** @var array<string, string> */
+    private array $parameters;
+
+    /**
+     * @param array<mixed> $value
+     */
+    public function __construct(array $value, Arguments $arguments)
+    {
+        $keys = array_filter(array_keys($value), fn ($key) => ! $arguments->has($key));
```
[the remark this answers is not in the document]
Comment (romm) [reply src/Mapper/Object/Exception/UnexpectedArrayKeysForClass.php]: Good catch!

File: src/Mapper/Object/FilledArguments.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        return $this->value[$name];
+    }
+
+    /**
+     * @param mixed $source
+     * @return mixed[]
+     */
+    private function transform($source): array
+    {
+        $argumentsNumber = count($this->arguments);
```
[the remark this answers is not in the document]
Comment (romm) [reply src/Mapper/Object/FilledArguments.php]: Right. I went for `$argumentsCount`.

**Comment by romm** [comment]
Hi @eigan, just fixed this issue, thanks for the report! Can you confirm it is ok by your side?

**Comment by eigan** [comment]
@romm Seems to be fixed 👍🏻 

```php
final class Foo
{
    public ?Bar $foo;
}

class Bar {
    public string $note = "hello";
}

$foo = (new \CuyZ\Valinor\MapperBuilder())
    ->mapper()
    ->map(Foo::class, [
        'foo' => 10
    ]);

print_r($foo);
```

## Result
```
Foo Object
    [bar] => Bar Object
        [note] => 10 (string)
```
This is not really strict. Though the PR doesnt really promise full-strictness either.. 😃 
1) Note is cast from int to string
2) 'foo' actually expects `{note: string}` but apparently just sending something that can be cast to the one and only property of Bar is allowed.

**Comment by romm** [comment]
Hi @eigan, I can't reproduce your last report, and the result you get is quite surprising. Are you sure you did run it on this feature branch?

**Comment by eigan** [comment]
@romm Hey, sorry. You are right, I was on `master` 🤦🏻‍♂️ 

My example was also wrong. Source should have had `bar`, not `foo` (updated now).

**Comment by romm** [comment]
Thanks a lot for all the testing work, very much appreciated!

I'll probably publish a new release with this feature tomorrow.

**Pull request merged**